src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -199,89 +366,114 @@ public ReadOnlySequence<T> Slice(long start, SequencePosition end)
         /// <param name="length">The length of the slice</param>
         public ReadOnlySequence<T> Slice(SequencePosition start, long length)
         {
-            BoundsCheck(start, _sequenceEnd); // check start before length
+            // Check start before length
+            uint sliceStartIndex = (uint)GetIndex(start);
+            object sliceStartObject = start.GetObject();
+
+            uint startIndex = (uint)GetIndex(_sequenceStart);
+            object startObject = _sequenceStart.GetObject();
+
+            uint endIndex = (uint)GetIndex(_sequenceEnd);
+            object endObject = _sequenceEnd.GetObject();
+
+            // Single-Segment Sequence
+            if (startObject == endObject)
+            {
+                if (!InRange(sliceStartIndex, startIndex, endIndex))
+                {
+                    ThrowHelper.ThrowArgumentOutOfRangeException_PositionOutOfRange();
+                }
+
+                if (length < 0)
+                    // Passing value >= 0 means throw exception on length argument
+                    ThrowHelper.ThrowStartOrEndArgumentValidationException(0);
+
+                if (endIndex - sliceStartIndex < length)
+                    ThrowHelper.ThrowArgumentOutOfRangeException_OffsetOutOfRange();
+
+                goto FoundInFirstSegment;
+            }
+
+            // Multi-Segment Sequence
+            var sliceStartSegment = (ReadOnlySequenceSegment<T>)sliceStartObject;
+            ulong sliceRange = (ulong)((sliceStartSegment.RunningIndex + sliceStartIndex));
+            ulong startRange = (ulong)(((ReadOnlySequenceSegment<T>)startObject).RunningIndex + startIndex);
+            ulong endRange = (ulong)(((ReadOnlySequenceSegment<T>)endObject).RunningIndex + endIndex);
+
+            // This optimization works because we know sliceStartIndex, startIndex, and endIndex are all >= 0
+            Debug.Assert(sliceStartIndex >= 0 && startIndex >= 0 && endIndex >= 0);
+            if (!InRange(sliceRange, startRange, endRange))
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_PositionOutOfRange();
+            }
+
             if (length < 0)
                 // Passing value >= 0 means throw exception on length argument
                 ThrowHelper.ThrowStartOrEndArgumentValidationException(0);
 
-            SequencePosition end = Seek(start, _sequenceEnd, length);
-            return SliceImpl(start, end);
+            if (sliceRange + (ulong)length > endRange)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_OffsetOutOfRange();
+            }
+
+            int currentLength = sliceStartSegment.Memory.Length - (int)sliceStartIndex;
```

Comment:
Add `if (currentLength < 0) ThrowHelper.ThrowArgumentOutOfRangeException_PositionOutOfRange();`

Reply:
Also `if (currentLength < 0) ThrowHelper.ThrowArgumentOutOfRangeException_PositionOutOfRange();` can be moved down inside `if (currentLength <= length)` before `goto FoundInFirstSegment;`